File: day13/AOC.py
import re
import networkx as nx
from itertools import permutations

def parse_input(input):
    graph = nx.DiGraph()
    for line in input:
        x = line.split()
        p1 = x[0]
        p2 = x[-1][:-1] # remove the period
        weight = int(x[3])
        if x[2] == "lose":
            weight = -weight
        graph.add_edge(p1, p2, weight=weight)
    return graph

def part1(input, me=False):
    graph = parse_input(input)
    if me:
        graph.add_node("me")
        for node in graph.nodes():
            if node != "me":
                graph.add_edge("me", node, weight=0)
                graph.add_edge(node, "me", weight=0)
                
    happiness = []
    for path in permutations(graph.nodes()):
        happy = 0
        for i in range(len(path) - 1):
            happy += graph.get_edge_data(path[i], path[(i+1)])['weight']
            happy += graph.get_edge_data(path[(i+1)], path[i])['weight']
        happy += graph.get_edge_data(path[0], path[-1])['weight']
        happy += graph.get_edge_data(path[-1], path[0])['weight']
        happiness.append(happy)
    return max(happiness)

def part2(input):
    return part1(input, me=True)

def main():
    input = open("input.txt", "r").read().splitlines()
    part1_sum = part1(input)
    print(f"Part 1: {part1_sum}")
    part2_sum = part2(input)
    print(f"Part 2: {part2_sum}")

if __name__ == "__main__":
    main()

File: Go/day13/AOC.go
package main

import (
	"fmt"
	"os"
	"strings"
)

type node struct {
	name      string
	neighbors map[string]int
}

func (n *node) addNeighbor(person string, happiness int) {
	if n.neighbors == nil {
		n.neighbors = make(map[string]int)
	}
	n.neighbors[person] = happiness
}

func parseInput(input []string) map[string]*node {
	nodes := make(map[string]*node)

	for _, line := range input {
		var name1, name2 string
		var op string
		var happiness int
		fmt.Sscanf(line, "%s would %s %d happiness units by sitting next to %s.", &name1, &op, &happiness, &name2)

		if op == "lose" {
			happiness = -happiness
		}

		if name2[len(name2)-1] == '.' {
			name2 = name2[:len(name2)-1]
		}

		if _, ok := nodes[name1]; !ok {

			nodes[name1] = &node{name: name1}
		}
		if _, ok := nodes[name2]; !ok {
			nodes[name2] = &node{name: name2}
		}

		nodes[name1].addNeighbor(name2, happiness)
	}

	return nodes
}

func generatePermutations(elements []string) [][]string {
	var result [][]string

	var generate func(int)
	generate = func(n int) {
		if n == 1 {
			temp := make([]string, len(elements))
			copy(temp, elements)
			result = append(result, temp)
			return
		}

		for i := 0; i < n; i++ {
			elements[i], elements[n-1] = elements[n-1], elements[i]
			generate(n - 1)
			elements[i], elements[n-1] = elements[n-1], elements[i]
		}
	}

	generate(len(elements))
	return result
}

func generateHappiness(nodes map[string]*node) int {
	var max int
	people := make([]string, 0, len(nodes))
	for name := range nodes {
		people = append(people, name)
	}

	permutations := generatePermutations(people)

	for _, perm := range permutations {
		mood := 0
		for i := 0; i < len(perm); i++ {
			mood += nodes[perm[i]].neighbors[perm[(i+1)%len(perm)]]
			if i == 0 {
				mood += nodes[perm[i]].neighbors[perm[len(perm)-1]]
			} else {
				mood += nodes[perm[i]].neighbors[perm[(i-1)]]
			}
		}
		if mood > max {
			max = mood
		}
	}

	return max
}

func part1(input []string) int {
	return generateHappiness(parseInput(input))
}

func part2(input []string) int {
	nodes := parseInput(input)
	nodes["me"] = &node{name: "me"}
	for name := range nodes {
		nodes[name].addNeighbor("me", 0)
		nodes["me"].addNeighbor(name, 0)
	}

	return generateHappiness(nodes)
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := strings.Split(string(input), "\n")
	part1_ans := part1(lines)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(lines)
	fmt.Printf("Part 2: %d\n", part2_ans)
}
